helix-view/src/clipboard/config.rs

```diff
@@ -0,0 +1,248 @@
+                paste: "xsel", "-i", "-b";
+                primary_copy: "xsel", "-o";
+                primary_paste: "xsel", "-i";
+                env: "DISPLAY";
+                test: "xsel", "-o", "-b";
+            }
+
+            "lemonade" {
+                copy: "lemonade", "copy";
+                paste: "lemonade", "paste";
+            }
+
+            #[cfg(unix)]
+            "doit" {
+                copy: "doitclient", "wclip";
+                paste: "doitclient", "wclip", "-r";
+            }
+
+            #[cfg(any(windows, target_os = "linux"))] // this is a godsend on WSL
+            "win32yank" {
+                copy: "win32yank.exe", "-i";
+                paste: "win32yank.exe", "-o";
+            }
+
+            #[cfg(unix)]
+            "tmux" {
+                copy: "tmux", "load-buffer", "-";
+                paste: "tmux", "save-buffer", "-";
+                env: "TMUX";
+            }
+
+            #[cfg(target_os = "linux")]
+            "termux" {
+                copy: "termux-clipboard-set";
+                paste: "termux-clipboard-get";
+            }
+
+            #[cfg(target_os = "macos")]
+            "pbcopy+pbpaste" {
+                copy: "pbcopy";
+                paste: "pbpaste";
+            }
+        };
+        if cfg!(windows) {
+            cfg.order.push("clipboard-win".to_owned());
+        }
+        cfg
+    }
+}
+
+impl ClipboardConfig {
+    pub fn get_provider(&self) -> Box<dyn ClipboardProvider> {
+        for name in &self.order {
+            #[cfg(windows)]
+            if name == "clipboard-win" {
+                return Box::new(provider::WindowsProvider);
+            }
+            if let Some(spec) = self.providers.get(name) {
+                if spec.is_available() {
+                    return Box::new(provider::CommandProvider::from(spec));
+                }
+            } else {
+                warn!("No clipboard provider named {name}");
+                continue;
+            }
+        }
+
+        Box::new(provider::NopProvider::new())
+    }
+}
+
+impl ClipboardProviderSpec {
+    pub fn is_available(&self) -> bool {
+        macro_rules! check {
+            ($cond:expr) => {
+                if !$cond {
+                    return false;
+                }
+            };
+        }
+
+        check!(!self.copy.is_empty() && exists(&self.copy[0]));
+        check!(!self.paste.is_empty() && exists(&self.paste[0]));
+
+        if let Some(copy) = &self.primary_copy {
+            check!(!copy.is_empty() && exists(&copy[0]));
+        }
+        if let Some(paste) = &self.primary_paste {
+            check!(!paste.is_empty() && exists(&paste[0]));
+        }
+        for var in &self.env_vars {
+            check!(env_var_is_set(var));
+        }
+        for cmd in &self.test_commands {
+            // FIXME: check performance of is_exit_success
+            check!(!cmd.is_empty() && is_exit_success(&cmd[0], &cmd[1..]))
+        }
```

Comment:
Isn't this check sufficient?

Reply:
If anything, *this* is the redundant one: the `test_commands` list is only used for `xsel`. In all other cases, it was deemed sufficient to check for the existence of the copy/paste binaries, and possibly an environment variable. For `xsel`, I think that might still be fine, especially seeing as it still has a `DISPLAY` check.